### PR TITLE
Fix CastOpSuite failures with Spark 3.4

### PR DESCRIPTION
**Pull request by andygrove**
Closes https://github.com/NVIDIA/spark-rapids/issues/7752

Disables the following casts in Spark 3.4 when ansiEnabled is true

- [x] BOOLEAN -> TIMESTAMP
- [x] DATE / TIMESTAMP -> BOOLEAN

Disables the following casts in Spark 3,4 regardless of ansiEnabled.

- [x] integral -> DATE
- [x] fp -> DATE

### CODE REVIEW

File: sql-plugin/src/main/scala/com/nvidia/spark/rapids/TypeChecks.scala [hunk 1]
```diff
@@ -1362,11 +1364,15 @@ class CastChecks extends ExprChecks {
...
+    GpuTypeShims.additionalTypesDateCanCastTo
+  def sparkDateSig: TypeSig = DECIMAL_128 + TIMESTAMP + DATE + STRING +
+    GpuTypeShims.additionalTypesDateCanCastTo
 
-  val timestampChecks: TypeSig = integral + fp + BOOLEAN + TIMESTAMP + DATE + STRING
-  val sparkTimestampSig: TypeSig = cpuNumeric + BOOLEAN + TIMESTAMP + DATE + STRING
+  def timestampChecks: TypeSig = integral + fp + TIMESTAMP + DATE + STRING +
+    GpuTypeShims.additionalTypesTimestampCanCastTo
+  def sparkTimestampSig: TypeSig = cpuNumeric + TIMESTAMP + DATE + STRING +
+    GpuTypeShims.additionalTypesTimestampCanCastTo
```
Comment (andygrove) [inline sql-plugin/src/main/scala/com/nvidia/spark/rapids/TypeChecks.scala:1375]: Moved `BOOLEAN` to `GpuTypeShims`

File: sql-plugin/src/main/scala/com/nvidia/spark/rapids/TypeChecks.scala [hunk 2]
```diff
@@ -1362,11 +1364,15 @@ class CastChecks extends ExprChecks {
       .withPsNote(TypeEnum.STRING, fpToStringPsNote)
   val sparkFpSig: TypeSig = cpuNumeric + BOOLEAN + TIMESTAMP + STRING
 
-  val dateChecks: TypeSig = integral + fp + BOOLEAN + TIMESTAMP + DATE + STRING
-  val sparkDateSig: TypeSig = cpuNumeric + BOOLEAN + TIMESTAMP + DATE + STRING
+  def dateChecks: TypeSig = TIMESTAMP + DATE + STRING +
+    GpuTypeShims.additionalTypesDateCanCastTo
+  def sparkDateSig: TypeSig = DECIMAL_128 + TIMESTAMP + DATE + STRING +
+    GpuTypeShims.additionalTypesDateCanCastTo
```
Comment (andygrove) [inline sql-plugin/src/main/scala/com/nvidia/spark/rapids/TypeChecks.scala:1370]: Moved `integral + fp` to `GpuTypeShims`

File: sql-plugin/src/main/scala/com/nvidia/spark/rapids/TypeChecks.scala [hunk 3]
```diff
@@ -1348,8 +1348,10 @@ class CastChecks extends ExprChecks {
     NULL + DECIMAL_128
   val sparkNullSig: TypeSig = all
 
-  val booleanChecks: TypeSig = integral + fp + BOOLEAN + TIMESTAMP + STRING + DECIMAL_128
-  val sparkBooleanSig: TypeSig = cpuNumeric + BOOLEAN + TIMESTAMP + STRING
+  def booleanChecks: TypeSig = integral + fp + BOOLEAN + STRING + DECIMAL_128 +
+    GpuTypeShims.additionalTypesBooleanCanCastTo
+  def sparkBooleanSig: TypeSig = cpuNumeric + BOOLEAN + STRING +
+    GpuTypeShims.additionalTypesBooleanCanCastTo
```
Comment (andygrove) [inline sql-plugin/src/main/scala/com/nvidia/spark/rapids/TypeChecks.scala:1354]: Moved `TIMESTAMP` to `GpuTypeShims`

File: sql-plugin/src/main/spark340/scala/com/nvidia/spark/rapids/shims/GpuTypeShims.scala [hunk 4]
```diff
@@ -0,0 +1,270 @@
...
+    TypeSig.none
+  } else {
+    TypeSig.BOOLEAN
+  }
+
+  def additionalTypesTimestampCanCastTo: TypeSig = if (SQLConf.get.ansiEnabled) {
+    TypeSig.none
+  } else {
+    TypeSig.BOOLEAN
+  }
```
Comment (andygrove) [inline sql-plugin/src/main/spark340/scala/com/nvidia/spark/rapids/shims/GpuTypeShims.scala:241]: This code is new in this PR. The rest of this file is copied from the 330 shim version.

**Comment by andygrove** [comment]
build

**Comment by andygrove** [comment]
build

File: sql-plugin/src/main/spark340/scala/com/nvidia/spark/rapids/shims/GpuTypeShims.scala (hunk 4 above)
Comment (andygrove) [reply sql-plugin/src/main/spark340/scala/com/nvidia/spark/rapids/shims/GpuTypeShims.scala:241]: Thanks. That makes sense.

**Comment by andygrove** [comment]
build

File: sql-plugin/src/main/spark340/scala/com/nvidia/spark/rapids/shims/GpuTypeShims.scala (hunk 4 above)
Comment (andygrove) [reply sql-plugin/src/main/spark340/scala/com/nvidia/spark/rapids/shims/GpuTypeShims.scala]: I have pushed new commits to resolve this.

File: sql-plugin/src/main/spark330/scala/com/nvidia/spark/rapids/shims/GpuTypeShims.scala [hunk 5]
```diff
@@ -227,6 +227,12 @@ object GpuTypeShims {
   def typesYearMonthCanCastToOnSpark: TypeSig = TypeSig.YEARMONTH + TypeSig.STRING +
       TypeSig.integral
 
+  def additionalTypesBooleanCanCastTo: TypeSig = TypeSig.TIMESTAMP
+
+  def additionalTypesDateCanCastTo:TypeSig = TypeSig.BOOLEAN + TypeSig.integral + TypeSig.fp
+
+  def additionalTypesTimestampCanCastTo:TypeSig = TypeSig.BOOLEAN
+
```
[the remark this answers is not in the document]
Comment (andygrove) [reply sql-plugin/src/main/spark330/scala/com/nvidia/spark/rapids/shims/GpuTypeShims.scala]: Thanks. Fixed.

**Comment by andygrove** [comment]
build

**Comment by andygrove** [comment]
build

**Review by gerashegalov — APPROVED** [review]
LGTM

**Comment by andygrove** [comment]
build

**Review by gerashegalov — APPROVED** [review]

**Pull request merged**